### PR TITLE
Minor Changes to DMRG(Block2)

### PR DESCRIPTION
**Pull request by lcyyork**
## Description
Some minor changes regarding to the new block2 interface.
- [x] Fix the `read_active_wfn` bug when multiple states are available.
- [x] Remove the MPS compressed intermediates for DSRG after the computation.
- [x] Fix the DMRG-SCF maxiter in the "nail-down" stage.
- [x] Read MPS compressed intermediates for LDSRG(2) computations.
- [x] Add Cholesky active orbitals for a trade-off of localizing orbitals while maintaining point-group symmetry. [(Fast noniterative orbital localization for large molecules)](https://pubs.aip.org/jcp/article/125/17/174101/984430/Fast-noniterative-orbital-localization-for-large)

## User Notes
- [x] Add Cholesky orbitals for localizing active orbitals.
- [x] Several options are added:
  - `BLOCK2_INTERGRAL_CUTOFF`: considered as zero if integrals are smaller than this threshold.
  - `BLOCK2_INITIAL_GUESS_OCC`: active occupation numbers (in Pitzer order) for DMRG initial guess. By default, we use random initial MPS. If not, one can specify this option in combination with `BLOCK2_INITIAL_GUESS`. My preliminary tests show the occupation bias (i.e., `BLOCK2_INITIAL_GUESS`) should be around 0.5.
  - `DSRG_3RDM_BLOCK2_CPS_BOND_DIMENSION`: bond dimension used to compute `Ψ(N-1) = h_{pσ} (t) |Ψ (N)> = \sum_{uvw} t_{pw}^{uv} \sum_{σ1} w^+_{σ1} v_{σ1} u_{σ} |Ψ(N)>`. My preliminary tests show this value should be at least three times the bond dimension of `|Ψ(N)>` to reach mEh accuracy for the 3-cumulant contribution in DSRG-MRPT2.

## Checklist
- [x] Added/updated tests of new features and included a reference `output.ref` file
- [x] Removed comments in code and input files
- [x] Documented source code
- [x] Checked for redundant headers
- [x] Checked for consistency in the formatting of the output file
- [x] Documented new features in the manual
- [x] Ready to go!

### CODE REVIEW

**Comment by fevangelista** [comment]
I'd like @brianz98 to check this PR since he has also done work on the DMRG testing.

File: forte/dmrg/block2_dmrg_solver.cc [hunk 1]
```diff
@@ -737,7 +737,22 @@ Block2DMRGSolver::compute_complementary_H2caa_overlap(const std::vector<size_t>&
...
+    int bond_dim = 500;
+    auto sweep_bond_dims = dmrg_options_->get_int_list("BLOCK2_SWEEP_BOND_DIMS");
+    if (sweep_bond_dims.size() != 0)
+        bond_dim = sweep_bond_dims.back();
+    std::vector<block2::ubond_t> ket0_bond_dims(1, bond_dim);
+    std::vector<block2::ubond_t> bra_bond_dims(1, 2 * bond_dim);
+    std::vector<double> noises{0.0};
+
     // system initialization
+    auto integral_cutoff = 1.0e-14;
```
Comment (brianz98) [inline forte/dmrg/block2_dmrg_solver.cc:755]: should this be hardcoded or passed in as option?

File: forte/dmrg/block2_dmrg_solver.cc [hunk 2]
```diff
@@ -192,7 +192,7 @@ struct Block2ScratchManager {
         // TODO may cause problem for MPI
         const std::vector<std::string> file_name_patterns = std::vector<std::string>{
             ".NPDM.FRAG.", ".MPS.",         ".MPS.INFO.", ".MMPS.", ".MMPS.INFO.",
-            ".MMPS-WFN.",  "-mps_info.bin", "@TMP.",      "@TMP-"};
+            ".MMPS-WFN.",  "-mps_info.bin", "@TMP.",      "@TMP-",  "DSRG-"};
```
Comment (brianz98) [inline forte/dmrg/block2_dmrg_solver.cc:195]: many of these intermediate scratch files are not documented - would appreciate it if you could  add some comment to explain a little bit what these files are here

**Review by brianz98 — COMMENTED** [review]
Thanks for this PR! I've left some comments

File: forte/dmrg/block2_dmrg_solver.cc [hunk 3]
```diff
@@ -737,7 +737,22 @@ Block2DMRGSolver::compute_complementary_H2caa_overlap(const std::vector<size_t>&
     const std::vector<int> tshape{na1, na1, na1};
     const std::vector<size_t> tstride{nactv * nactv, 1, nactv};
 
+    // reset stack memory
+    auto stack_mem =
+        static_cast<size_t>(dmrg_options_->get_double("BLOCK2_STACK_MEM") * 1024 * 1024 * 1024);
+    impl_->stack_mem_ = stack_mem;
```
Comment (hczhai) [inline forte/dmrg/block2_dmrg_solver.cc:743]: These four lines (L740-L743) for resetting memory may be deleted.

**Comment by lcyyork** [comment]
> I'd like @brianz98 to check this PR since he has also done work on the DMRG testing. It also looks like there is some cleanup that needs to be done. BTW, can you say more about the changes made in this PR? The nail-down stage code seems mostly commented out, IIRC.

I am sorry, but this PR is not ready for review.

File: forte/mcscf/mcscf_orb_grad.cc [hunk 4]
```diff
@@ -763,6 +764,34 @@ double MCSCF_ORB_GRAD::evaluate(std::shared_ptr<psi::Vector> x, std::shared_ptr<
     return energy_;
 }
 
+// std::shared_ptr<psi::Matrix> MCSCF_ORB_GRAD::localize_orbitals() {
```
[the remark this answers is not in the document]
Comment (lcyyork) [reply forte/mcscf/mcscf_orb_grad.cc:767]: The ordering localized orbitals need to be consistent for consecutive dmrg-ci calculations. I am not sure if there is a simple way to do this.

[the remark this answers is not in the document]
Comment (lcyyork) [reply forte/mcscf/mcscf_orb_grad.cc:767]: These lines are removed.

File: forte/dmrg/block2_dmrg_solver.cc (hunk 1 above)
Comment (lcyyork) [reply forte/dmrg/block2_dmrg_solver.cc]: An option is added.

File: forte/mcscf/mcscf_2step.cc [hunk 5]
```diff
@@ -330,47 +330,50 @@ double MCSCF_2STEP::compute_energy() {
...
-                        options_->set_int_list("DMRG_SWEEP_MAX_SWEEPS", {4, 4, 6});
-                        options_->set_double_list("DMRG_SWEEP_NOISE_PREFAC", {1.0e-2, 5.0e-3, 0.0});
-                        options_->set_double_list("DMRG_SWEEP_DVDSON_RTOL",
-                                                  {1.0e-5, 1.0e-6, 1.0e-7});
-                        options_->set_double_list("DMRG_SWEEP_ENERGY_CONV",
-                                                  {1.0e-6, 1.0e-7, 1.0e-8});
-                    }
-                } else {
-                    as_solver_->set_maxiter(++mci_maxiter_);
+                    //     mci_maxiter_ = 14;
```
[the remark this answers is not in the document]
Comment (lcyyork) [reply forte/mcscf/mcscf_2step.cc]: I have to go through a lot of trials and errors. These lines are restored.

**Review by hczhai — APPROVED** [review]
The DMRG part looks good to me. Thanks for the changes and tests!

**Review by brianz98 — CHANGES_REQUESTED** [review]
Looks good to me! Could you also add the new option(s) to the documentation?

**Comment by fevangelista** [comment]
Thanks everyone! Is this ready for merging now?

**Comment by lcyyork** [comment]
> Thanks everyone! Is this ready for merging now?

I think yes.

**Pull request merged**